tools/tco/tco.cpp

```diff
@@ -0,0 +1,119 @@
+//===- tco.cpp - Tilikum Crossing Opt ---------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This is to be like LLVM's opt program, only for FIR.  Such a program is
+// required for roundtrip testing, etc.
+//
+//===----------------------------------------------------------------------===//
+
+#include "flang/Optimizer/Dialect/FIRDialect.h"
+#include "flang/Optimizer/Support/KindMapping.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/Module.h"
+#include "mlir/InitAllDialects.h"
+#include "mlir/Parser.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/Passes.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/ErrorOr.h"
+#include "llvm/Support/InitLLVM.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/SourceMgr.h"
+#include "llvm/Support/ToolOutputFile.h"
+#include "llvm/Support/raw_ostream.h"
+
+namespace {
+
+using namespace llvm;
+
+cl::opt<std::string> inputFilename(cl::Positional, cl::desc("<input file>"),
+                                   cl::init("-"));
+
+cl::opt<std::string> outputFilename("o", cl::desc("Specify output filename"),
+                                    cl::value_desc("filename"), cl::init("-"));
+
+cl::opt<bool> emitFir("emit-fir", cl::desc("Parse and pretty-print the input"),
+                      cl::init(false));
+
+void printModuleBody(mlir::ModuleOp mod, raw_ostream &output) {
+  // don't output the terminator bogo-op
+  auto e{--mod.end()};
```

Comment:
Note: Block has a utility for this `Block::without_terminator`.

Reply:
Thanks. I was looking for that. :)